File: flutter_app/lib/models/fitness_log.dart
enum FitnessLogType { meal, workout }

class FitnessLogModel {
  final String id;
  final String userId;
  final FitnessLogType type;
  final String content;
  final int? calories;
  final Map<String, dynamic>? parsedData;
  final DateTime timestamp;

  FitnessLogModel({
    required this.id,
    required this.userId,
    required this.type,
    required this.content,
    this.calories,
    this.parsedData,
    required this.timestamp,
  });

  factory FitnessLogModel.fromJson(Map<String, dynamic> json) => FitnessLogModel(
        id: (json['log_id'] ?? json['id']) as String,
        userId: (json['user_id'] ?? json['userId']) as String,
        type: _typeFrom(json['log_type'] as String?),
        content: json['content'] as String,
        calories: json['calories'] as int?,
        parsedData: (json['ai_parsed_data'] ?? json['parsedData']) is Map ?
            ((json['ai_parsed_data'] ?? json['parsedData']) as Map).cast<String, dynamic>() : null,
        timestamp: DateTime.parse(json['timestamp'] as String),
      );

  Map<String, dynamic> toJson() => {
        'log_id': id,
        'user_id': userId,
        'log_type': type.name,
        'content': content,
        'calories': calories,
        'ai_parsed_data': parsedData,
        'timestamp': timestamp.toIso8601String(),
      }..removeWhere((k, v) => v == null);

  static FitnessLogType _typeFrom(String? v) {
    switch ((v ?? 'meal').toLowerCase()) {
      case 'workout':
        return FitnessLogType.workout;
      default:
        return FitnessLogType.meal;
    }
  }
}

File: app/models/fitness_log.py
from __future__ import annotations

from datetime import datetime, timezone
from enum import Enum
from typing import Any, Dict, Optional
from uuid import uuid4

from pydantic import BaseModel, Field, field_validator


def _now_utc() -> datetime:
    return datetime.now(timezone.utc)


class FitnessLogType(str, Enum):
    meal = "meal"
    workout = "workout"
    water = "water"
    supplement = "supplement"


class FitnessLog(BaseModel):
    log_id: str = Field(default_factory=lambda: str(uuid4()))
    user_id: str
    log_type: FitnessLogType
    content: str
    calories: Optional[int] = None
    timestamp: datetime = Field(default_factory=_now_utc)
    ai_parsed_data: Dict[str, Any] = Field(default_factory=dict)
    client_generated_id: Optional[str] = None  # 🔑 For optimistic UI matching

    @field_validator("timestamp")
    @classmethod
    def _ensure_timezone(cls, value: datetime) -> datetime:
        if value.tzinfo is None:
            return value.replace(tzinfo=timezone.utc)
        return value.astimezone(timezone.utc)

    def to_dict(self) -> Dict[str, Any]:
        return self.model_dump(mode="python", exclude_none=True)

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> "FitnessLog":
        # Convert Firestore datetime objects to ISO strings
        if "timestamp" in data and hasattr(data["timestamp"], "isoformat"):
            data["timestamp"] = data["timestamp"].isoformat()
        return cls(**data)



